pymongo/pool.py

```diff
@@ -1313,9 +1368,26 @@ def _get_socket(self, all_credentials):
             deadline = _time() + self.opts.wait_queue_timeout
         else:
             deadline = None
-        if not self._socket_semaphore.acquire(
-                True, self.opts.wait_queue_timeout):
-            self._raise_wait_queue_timeout()
+
+        with self.size_cond:
+            self._raise_if_not_ready()
+            if self.waiters >= self.max_waiters:
```

Comment:
Because we don't need to increment the # of waiters to know that there are already too many.

Reply:
Because we don't need to increment the # of waiters to know that there are already too many. Incrementing first means we also need to decrement after raising the error so it seemed more straightforward to check first, then increment.